devito/ir/clusters/cluster.py

```diff
@@ -163,14 +163,44 @@ def is_scalar(self):
     @cached_property
     def is_dense(self):
         """
-        True if the Cluster unconditionally writes into DiscreteFunctions
-        through affine access functions, False otherwise.
+        A Cluster is dense if at least one of the following conditions is True:
+
+            * It is defined over a unique Grid and all of the Grid Dimensions
+              are PARALLEL.
+            * Only DiscreteFunctions are written and only affine index functions
+              are used (e.g., `a[x+1, y-2]` is OK, while `a[b[x], y-2]` is not)
         """
+        # Hopefully it's got a unique Grid and all Dimensions are PARALLEL.
+        # This is a quick and easy check so we try it first
+        try:
+            grid = self.grid
+            for d in grid.dimensions:
+                if not any(PARALLEL in v for k, v in self.properties.items()
+                           if d in k._defines):
+                    raise ValueError
+            return True
+        except ValueError:
+            pass
+
+        # Fallback to legacy is_dense checks
         return (not any(e.conditionals for e in self.exprs) and
                 not any(f.is_SparseFunction for f in self.functions) and
                 not self.is_scalar and
                 all(a.is_regular for a in self.scope.accesses))
 
+    @cached_property
+    def grid(self):
+        if len(self.grids) == 1:
+            return self.grids[0]
+        raise ValueError("Cluster has no unique Grid")
+
+    @cached_property
+    def grids(self):
+        """
+        The Grid's over which the Cluster is defined.
```

Comment:
Grids' ?

Reply:
I've been told that one should use an apostrophe for the plural of proper nouns... :grimacing: